eyelid-match-ops/src/primitives/poly/modular_poly/modulus.rs

```diff
@@ -86,7 +85,28 @@ pub fn mod_poly_ark_ref(dividend: &Poly) -> Poly {
 ///
 /// This uses an [`ark-poly`] library implementation, and entirely replaces the inner polynomial representation.
 #[cfg(inefficient)]
-pub fn mod_poly_ark_mut(dividend: &mut Poly) {
+pub fn mod_poly_ark_mut<const MAX_POLY_DEGREE: usize>(dividend: &mut Poly<MAX_POLY_DEGREE>) {
     let remainder = mod_poly_ark_ref(dividend);
     *dividend = remainder;
 }
+
+/// Constructs and returns a new polynomial modulus used for the polynomial field, `X^[MAX_POLY_DEGREE] + 1`.
+/// This means that `X^[MAX_POLY_DEGREE] = -1`.
+///
+/// This is the canonical but un-reduced form of the modulus, because the reduced form is the zero polynomial.
+///
+/// TODO: work out how to generically make this a lazy static.
+/// Crates like `interned` or `lazy_static` might help, but we'll have to expand their macros and make them generic.
+pub fn new_unreduced_poly_modulus_slow<const MAX_POLY_DEGREE: usize>() -> Poly<MAX_POLY_DEGREE> {
+    let mut poly = Poly::zero();
+
+    // Since the leading coefficient is non-zero, this is in canonical form.
+    // Resize to the maximum size first, to avoid repeated reallocations.
+    poly[MAX_POLY_DEGREE] = Coeff::one();
+    poly[0] = Coeff::one();
+
+    // Check canonicity and degree.
+    assert_eq!(poly.degree(), MAX_POLY_DEGREE);
```

Comment:
My understanding is:
- The modulus has degree `MAX_POLY_DEGREE`.
- Reduced polynomials have degree `MAX_POLY_DEGREE - 1`

Is that correct and intended?

Reply:
Because an alternative could be:
- The modulus is actually `Poly<MAX_POLY_DEGREE + 1>`
- Reduced polynomials are `Poly<MAX_POLY_DEGREE>`

Reply:
Polynomials with degree `MAX_POLY_DEGREE - 1` have `MAX_POLY_DEGREE` coefficients, because the coefficients have degrees `0, 1, ... , MAX_POLY_DEGREE - 1`.

Having an even number of coefficients is a requirement for efficient Karatsuba implementations. Otherwise we have to pad the polynomial with the odd number of coefficients to the next power of two when we split them.

So currently we:
* declare `MAX_POLY_DEGREE = 2048`
* have a modulus with degree `MAX_POLY_DEGREE = 2048` and 2049 coefficients
* have polynomials with degree `MAX_POLY_DEGREE - 1 = 2047` and 2048 coefficients (which is a performance requirement)

This is slightly confusing, and could benefit from some renames and calculated constants. For example:
https://github.com/Inversed-Tech/eyelid/pull/47/files/a6821dcd3aa5a11af24e57ce60b872babd3eeadc#r1571498268

Under the alternative scheme, we would:
* declare `MAX_POLY_DEGREE = 2047`
* have a modulus with degree `MAX_POLY_DEGREE + 1 = 2048` and 2049 coefficients
* have polynomials with degree `MAX_POLY_DEGREE = 2047` and 2048 coefficients

The underlying sizes wouldn't change, but this way of coding them seems to be more confusing, and could cause more off-by-one errors.

Reply:
Yes, it is confusing, thanks for putting it together. I think it would be better to use `MAX_POLY_DEGREE = 2047` and create aliases for maximal number of coefficients, and the degree of the modulus.